grails-app/services/com/unifina/service/PermissionService.groovy

```diff
@@ -265,17 +268,46 @@ class PermissionService {
      * @return granted permission
      */
 	Permission systemGrant(Userish target, resource, Operation operation=Operation.READ) {
+		return systemGrant(target, resource, operation, null, null)
+	}
+
+	Permission systemGrant(Userish target, resource, Operation operation=Operation.READ, Subscription subscription, Date endsAt) {
 		target = target.resolveToUserish()
 		String userProp = getUserPropertyName(target)
 		String resourceProp = getResourcePropertyName(resource)
 
+		if (userProp == "user" && resourceProp == "stream" && operation == Operation.READ) {
```

Comment:
The `grantInboxStreamPermissions()` method anyway calls methods in the `StreamService` using the bean. So could we simply have this whole method in the `StreamService` and call it in the same way using:
```
StreamService streamService = grailsApplication.mainContext.getBean(StreamService)
```

Reply:
This ties back to @harbu 's [comment](https://github.com/streamr-dev/engine-and-editor/pull/583#discussion_r279562061). For me, the current proposal actually made it sort of worse:

Original situation:
- Inbox-specific stuff in `PermissionService`

Current situation
- Inbox-specific stuff in `PermissionService`
- Circular dependency between `StreamService` and `PermissionService`
- Creation of `Permission`s in `StreamService` (wrong place)

If these can't be cleanly decoupled via some event system as suggested above, then it's probably best to just go with the original problem of "Inbox-specific stuff in `PermissionService`". Maybe add a comment there to try and fix it later in Grails 3.

Reply:
There is actually something called [GORM events](https://grails.github.io/grails2-doc/2.5.x/guide/single.html#eventsAutoTimestamping). We could tap into the `afterInsert`/`beforeDelete` events for `Permission` objects to create/delete the child permissions. That won't be 100% decoupled either as we'd define the handlers (which are inbox-stream-specific) in `Permission.groovy`, but it seems cleaner than the current impl.

Reply:
I think the original solution (at least the version which I read) also had the circular dependency problem. It was under a private helper method.

In one way the original solution is the most cleanest if you think about it from the perspective of someone trying to understand / debug exactly what the method `systemGrant` does. There is less jumping around code and what the method does is laid down quite straightforward.

The perspective I had was from someone reading the code from a mindset of "how do inbox streams work?". Having all inbox stream functionality lumped into one class does aid in that. However, it does hamper readability of someone trying to understand what exactly `systemGrant` does because there is more jumping between methods involved. 

The event emitting solution has its own pros/cons. It decouples the PermissionService from inbox stream functionality nicely. What feels a bit off about it is the inbox stream functionality being defined in the `Permission.groovy` class. I think it could potentially make it hard to find / surprise a developer with its behavior. Another concern I have is unit tests: are events emitted correctly in unit tests? (GORM support in unit tests is known to be rather spotty) . The latter concern can be tested quickly ofc.

Maybe a more flexible event system would work best, but I don't think implementing a custom Observer pattern on PermissionService makes sense in this one case.

Reply:
I think the original solution is probably the best in this case tbh.